### PR TITLE
feat: submitted state in context 

### PR DESCRIPTION
**Pull request by macrozone**
see https://github.com/vazco/uniforms/issues/889

fixes https://github.com/vazco/uniforms/issues/889

### CODE REVIEW

File: packages/uniforms-bootstrap3/__tests__/_createContext.ts [hunk 1]
```diff
@@ -28,8 +29,8 @@ export default function createContext(
         placeholder: false,
         readOnly: false,
         showInlineError: false,
-        ...context?.state,
-      },
-    },
+        ...context?.state
+      }
+    }
```
Comment (macrozone) [inline packages/uniforms-bootstrap3/__tests__/_createContext.ts:34]: prettier config might have changed. it says `trailingComma: "none"`, so i guess the reformat is correct, but i am unsure

File: packages/uniforms/__tests__/BaseForm.tsx [hunk 2]
```diff
@@ -31,11 +31,11 @@ describe('BaseForm', () => {
 
   it('have correct context', () => {
     const wrapper = mount<BaseForm<any>>(
-      <BaseForm error={error} model={model} schema={schema} />,
+      <BaseForm error={error} model={model} schema={schema} />
     );
 
     const context = wrapper.instance().getContext();
-    expect(context).toEqual({
+    expect(context).toEqual<Context<any>>({
```
Comment (macrozone) [inline packages/uniforms/__tests__/BaseForm.tsx:38]: this way, the expected object is type checked. This helps to not forget properties

File: packages/uniforms/__tests__/BaseForm.tsx [hunk 3]
```diff
@@ -216,37 +218,54 @@ describe('BaseForm', () => {
...
       expect(wrapper.state('resetCount')).toBe(1);
     });
+    it('sets submitted back to false', async () => {
+      const wrapper = createWrapper();
+      const instance = wrapper.instance();
+      expect(instance.getContext().submitted).toBe(false);
+      wrapper.find('form').simulate('submit');
+      expect(instance.getContext().submitted).toBe(true);
+      instance.reset();
+      expect(instance.getContext().submitted).toBe(false);
```
Comment (macrozone) [inline packages/uniforms/__tests__/BaseForm.tsx:239]: i decided that reset will also reset this new property. I guess one would expect that

**Comment by macrozone** [comment]
@radekmie linting does not work locally, it says that `sh: eslint: command not found`, which makes sense as eslint is not in the devdependencies.

**Comment by wadamek65** [comment]
@macrozone `eslint` is not in the dev dependencies because it is a peer dependency of `eslint-config-vazco`. We are using `npm@7` which installs these for us automatically. Please make sure you are using `npm@7` when installing the dependencies and lint the code properly before submitting this for review.

**Comment by macrozone** [comment]
> @macrozone `eslint` is not in the dev dependencies because it is a peer dependency of `eslint-config-vazco`. We are using `npm@7` which installs these for us automatically. Please make sure you are using `npm@7` when installing the dependencies and lint the code properly before submitting this for review.

i have done this now

File: packages/uniforms-antd/__tests__/_createContext.ts [hunk 4]
```diff
@@ -19,6 +19,7 @@ export default function createContext(
       onSubmit() {},
       randomId,
       submitting: false,
+      submitted: false,
```
Comment (radekmie) [inline packages/uniforms-antd/__tests__/_createContext.ts:22]: Do keep these properties ordered.

File: packages/uniforms/__tests__/BaseForm.tsx [hunk 5]
```diff
@@ -216,33 +218,50 @@ describe('BaseForm', () => {
...
+        <BaseForm model={model} schema={schema} onSubmit={onSubmit} />,
+      );
 
     it('increase `resetCount`', () => {
+      const wrapper = createWrapper();
       wrapper.instance().reset();
 
       expect(wrapper.state('resetCount')).toBe(1);
     });
+    it('sets submitted back to false', async () => {
```
Comment (radekmie) [inline packages/uniforms/__tests__/BaseForm.tsx:232]:
```suggestion
    });

    it('sets submitted back to false', async () => {
```

File: packages/uniforms/__tests__/BaseForm.tsx [hunk 6]
```diff
@@ -256,7 +275,15 @@ describe('BaseForm', () => {
...
     });
 
+    it('sets `submitted` to true', async () => {
+      const wrapper = createWrapper();
+      const instance = wrapper.instance();
+      expect(instance.getContext().submitted).toBe(false);
+      wrapper.find('form').simulate('submit');
+      expect(instance.getContext().submitted).toBe(true);
+    });
     it('sets `submitting` state while submitting', async () => {
```
Comment (radekmie) [inline packages/uniforms/__tests__/BaseForm.tsx:285]:
```suggestion
    });

    it('sets `submitting` state while submitting', async () => {
```

File: packages/uniforms/src/BaseForm.tsx [hunk 7]
```diff
@@ -251,6 +255,7 @@ export class BaseForm<
       event.preventDefault();
       event.stopPropagation();
     }
+    this.setState({ submitted: true });
```
Comment (radekmie) [inline packages/uniforms/src/BaseForm.tsx:258]: It'd be better to skip unneeded updates:
```suggestion

    this.setState(state => state.submitted ? null : { submitted: true });
```

File: packages/uniforms/src/ValidatedForm.tsx [hunk 8]
```diff
@@ -119,7 +119,7 @@ export function Validated<Base extends typeof BaseForm>(Base: Base) {
         event.stopPropagation();
       }
 
-      this.setState({ validate: true });
+      this.setState({ validate: true, submitted: true });
```
Comment (radekmie) [inline packages/uniforms/src/ValidatedForm.tsx:122]:
```suggestion
      this.setState({ submitted: true, validate: true });
```

**Review by radekmie — CHANGES_REQUESTED** [review]
I've added some comments, most of them are code style-related. Also, please do not create unnecessary PRs - you could've continued on #890.

File: packages/uniforms/src/BaseForm.tsx (hunk 7 above)
Comment (macrozone) [reply packages/uniforms/src/BaseForm.tsx:258]: react already does that internally.

**Comment by macrozone** [comment]
i personally do not like that level of nitpicking, you should fix your linter if you want to enforce ordering rules of properties. Its bikeshedding after all. But your repo, your rules i gues...

File: packages/uniforms/src/BaseForm.tsx (hunk 7 above)
Comment (radekmie) [reply packages/uniforms/src/BaseForm.tsx]: No, it's not - here's a demo [CodeSandbox](https://codesandbox.io/s/practical-gareth-di8sb?file=/src/App.js). In class components you have to return `null`, in `useState` hook it works only if `newState === oldState`.

**Comment by macrozone** [comment]
i pushed the requested changes. 

There are two problems with the repo by the way:

- prettierconfig is wrong and removes trailing commas, but eslint complains about missing trailing commas. had to run `npm run lint:code -- --fix` to correct this.
- precommit hook (lint-staged) hanged forever (unclear why), so i canceled it, but then all my untaged changes where gone. Happend to me 3 times now.

**Review by radekmie — APPROVED** [review]

**Comment by macrozone** [comment]
hooray, thank you!

**Review by Monteth — APPROVED** [review]

**Pull request merged**